File: torchsl/csrc/multiview/mvpca.cpp
#include <vector>
#include <tuple>
#include <torch/script.h>
#include "multiview_helpers.h"

namespace torchsl {
std::tuple<torch::Tensor, torch::Tensor> mvpca(
    const std::vector<torch::Tensor>& Xs,
    const bool cross_covariance = true
    ) {
    const auto options = Xs[0].options().requires_grad(false);
    const int num_components = Xs[0].size(0);
    const auto dims = torchsl::dimensions(Xs);

    auto Xs_centered = torchsl::centerize(Xs);
    torch::Tensor cov;
    if (cross_covariance) {
        cov = torchsl::multiview_covariance_matrix(dims, [&Xs_centered] (int j, int r) {
            return Xs_centered[j].t().mm(Xs_centered[r]);
        }, options);
    } else {
        cov = torchsl::multiview_covariance_matrix(dims, [&Xs_centered] (int j, int r) {
            return (j == r) ? Xs_centered[j].t().mm(Xs_centered[r]) : torch::tensor(0);
        }, options);
    }
    cov /= num_components - 1;
    auto SI = torch::eye(dims.sum().item<int>(), options);

    return std::make_tuple(SI, cov);
}
}

File: torchsl/functional/multiview/mvpca.py
import torch
import torchsl

from .multiview_helpers import *

__all__ = ['mvpca']


# ===========================================================
# Multiview Principal Componant Analysis
# ===========================================================
def mvpca(Xs, cross_covariance=True, C=True):
    if cross_covariance and len(set(X.size(1) for X in Xs)) > 1:
        raise ValueError(f"MvPCA with cross covariance only works with multiview data with "
                         f"same dimensions, got dimensions of {tuple(X.size(1) for X in Xs)}")
    if C:
        return torchsl.ops.mvpca(Xs, cross_covariance)

    options = dict(dtype=Xs[0].dtype, device=Xs[0].device)
    num_components = Xs[0].size(0)
    dims = dimensions(Xs)

    Xs_centered = [X - X.mean(0) for X in Xs]
    if cross_covariance:
        cov = multiview_covariance_matrix(dims,
                                          lambda j, r: Xs_centered[j].t().mm(Xs_centered[r]),
                                          options)
    else:
        cov = multiview_covariance_matrix(dims,
                                          lambda j, r: Xs_centered[j].t().mm(Xs_centered[r])
                                          if j == r else 0,
                                          options)
    cov /= num_components - 1
    SI = torch.eye(sum(dims), **options)

    return SI, cov
